docs/source/jit.rst

```diff
@@ -203,6 +206,39 @@ Example::
   In our examples, we use comment-based annotations to ensure Python 2
   compatibility as well.
 
+Optional Type Refinement:
+
+TorchScript will refine the type of a variable of type Optional[T] when
+a comparison to None is made inside the conditional of an if statement.
+The compiler can reason about multiple None checks that are combined with
+AND, OR, or NOT. Refinement will also occur for else blocks of if statements
+that are not explicitly written.
+
+The expression must be emitted within the conditional; assigning
+a None check to a variable and using it in the conditional will not refine types.
+You can manually unwrap an optional with torch.jit._unwrap_optional(opt_var),
```

Comment:
It doesn't make sense to document an API that starts with an `_`. Either do not document it and leave it internal, or document it and remove the `_`.  For now I am in favor of leaving it out since type refinement is a nicer way of unwrapping the optional.

Reply:
Sounds good. Eventually if/when we support control flow & exceptions `assert x is not None` will be a nicer way of exposing the same functionality